### PR TITLE
STJ: Avoid duplicate initialization of required or init-only properties

### LINKED ISSUE #97621: Source generator in System.Text.Json does not handle shadowed members correctly
### Description

C# 12, source generator failed for 'new' or 'override' properties.

### Reproduction Steps

```C#
public record PlaceBase
{
    public string? Region { get; init; }
}
public record PlaceCommon : PlaceBase
{
    public new required string Region { get; init; }
}

[JsonSerializable(typeof(PlaceBase))]
[JsonSerializable(typeof(PlaceCommon))]
public partial class ApiModelJsonSerializerContext : JsonSerializerContext
{
}
```

### Expected behavior

PlaceCommon's 'Region' should be generated with its own Place property and ignore PlaceBase's 'Region' property.

### Actual behavior

Double 'Region' property generated with "Error CS1912 Duplicate initialization of member 'Region‘"

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by manandre**
Fixes #97621

### CODE REVIEW

File: src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs [hunk 1]
```diff
@@ -664,6 +664,27 @@ public class ClassWithCustomRequiredPropertyName
...
+        }
+
+        public class BaseClassWithInitOnlyProperty
+        {
+            public int MyInt { get; init; }
+        }
+
+        public class DerivedClassWithRequiredInitOnlyProperty : BaseClassWithInitOnlyProperty
+        {
+            public new required int MyInt { get; init; }
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs:685]: Would it be possible to also add test cases for `required` only or `init` only properties? They don't have to be on a separate test, you could just add more properties in that particular class.

**Review by eiriktsarpalis — COMMENTED** [review]
There's a test suite in the repo call `PropertyVisibilityTests`. As part of this change, would it be possible to evaluate the test coverage for shadowed properties (i.e. properties with the new keyword?)

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs [hunk 2]
```diff
@@ -1458,7 +1459,10 @@ private bool IsValidDataExtensionPropertyType(ITypeSymbol type)
                                 ParameterIndex = matchingConstructorParameter?.ParameterIndex ?? paramCount++,
                             };
 
-                            (propertyInitializers ??= new()).Add(propertyInitializer);
+                            if (propertyInitializers?.Any(p => propertyInitializer.Name.Equals(p.Name, GetStringComparison(options))) is not true)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:1462]: I suspect that the correct approach would be to deduplicate shadowed properties in the `ParsePropertyGenerationSpecs` method and not here.

File: src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs (hunk 1 above)
Comment (manandre) [reply src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs:690]: Test extended.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs (hunk 2 above)
Comment (manandre) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:1462]: I started from there but I was lost with the fast-path flow, in particular around the `AddPropertyWithConflictResolution` function. Can you please help me by clarifying the link between fast-path properties and property initializers?

**Comment by manandre** [comment]
> There's a test suite in the repo call `PropertyVisibilityTests`. As part of this change, would it be possible to evaluate the test coverage for shadowed properties (i.e. properties with the new keyword?)

Some tests around shadowed properties already exist but do not cover combination with required and init-only properties, what should be addressed by this PR.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:1462]: > between fast-path properties and property initializers?

One governs generation of fast-path serialization methods, while the other controls what properties should be instantiated using member-initializer syntax in the deserializer. Both are fed data deriving from the disambiguating logic in `AddPropertyWithConflictResolution`.

Comment (manandre) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs]: I have found a way to move the deduplication logic in the `ParsePropertyGenerationSpecs`. It looks way better like this.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs [hunk 3]
```diff
@@ -891,7 +891,9 @@ private TypeGenerationSpec ParseTypeGenerationSpec(in TypeToGenerate typeToGener
                             // property is static or an indexer
                             propertyInfo.IsStatic || propertyInfo.Parameters.Length > 0 ||
                             // It is overridden by a derived property
-                            PropertyIsOverriddenAndIgnored(propertyInfo, state.IgnoredMembers))
+                            PropertyIsOverriddenAndIgnored(propertyInfo, state.IgnoredMembers) ||
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:894]: There's an equivalent implementation in the reflection resolver:

https://github.com/dotnet/runtime/blob/3a382f2c51f02bc88869f4d05bfdb9d3c1f8e166/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultJsonTypeInfoResolver.Helpers.cs#L141

Should this be updated as well?

File: src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs [hunk 4]
```diff
@@ -664,6 +664,35 @@ public class ClassWithCustomRequiredPropertyName
             public required int PropertyWithInitOnlySetter { get; init; }
         }
 
+        [Fact]
+        public async Task DerivedClassWithRequiredProperty()
+        {
+            var value = new DerivedClassWithRequiredInitOnlyProperty { MyInt = 42, MyBool = true, MyLong = 4242 };
+            string json = await Serializer.SerializeWrapper(value);
+            Assert.Equal("""{"MyInt":42,"MyBool":true,"MyString":null,"MyLong":4242}""", json);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs:672]: Would it be possible to set a value to the `MyString` property to make sure it's being populated properly?

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs (hunk 3 above)
Comment (manandre) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:894]: Yes, we should keep both implementations aligned. For reflection flow, the deduplication is done at one place for both get and set operations, whereas for the source generation flow the fast-path and initializers deduplication logic were split. With this PR it is handled at the same place for all flows.

File: src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs [hunk 5]
```diff
@@ -664,6 +664,35 @@ public class ClassWithCustomRequiredPropertyName
...
+            public int MyInt { get; init; }
+            public bool MyBool { get; init; }
+            public string MyString { get; set; }
+        }
+
+        public class DerivedClassWithRequiredInitOnlyProperty : BaseClassWithInitOnlyProperty
+        {
+            public new required int MyInt { get; init; }
+            public new required bool MyBool { get; set; }
+            public new string MyString { get; init; }
```
[the remark this answers is not in the document]
Comment (manandre) [reply src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs:703]: Test cases added.

File: src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs (hunk 4 above)
Comment (manandre) [reply src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs]: Indeed.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs (hunk 3 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:894]: Taking a closer look at the implementation -- this method ultimately calls into the `AddPropertyWithConflictResolution` method which handles conflict resolution (including shadowed properties):

https://github.com/dotnet/runtime/blob/b94a82d63580775b2adb0ccac478719b8cd7049b/src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs#L1037-L1039

My impression is that the bug is caused by a missed case in that logic, so it seems to me that it's this particular logic that should be debugged. 

If you don't know how to debug source generator code, the easiest way to do that is by adding a unit test in `System.Text.Json.SourceGeneration.Roslyn*.Unit.Tests` and debugging them.

File: src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs [hunk 6]
```diff
@@ -664,6 +664,53 @@ public class ClassWithCustomRequiredPropertyName
...
+        {
+            public int MyInt { get; init; }
+            public bool MyBool { get; init; }
+            public string MyString { get; set; }
+            public string MyProp { get; init; }
+
+            public string MyMember1;
+            public string MyMember2 { get; init; }
+
+            public string MyField;
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/Common/RequiredKeywordTests.cs:696]: The base member must be defined as a member initializer to trigger the bug condition
```suggestion
            public required int MyField;
```

**Comment by eiriktsarpalis** [comment]
Good catch, I've reverted the changes.

File: src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/JsonSourceGeneratorTests.cs [hunk 7]
```diff
@@ -763,10 +763,39 @@ internal partial class JsonContext : JsonSerializerContext
                     {
                     }
                 }
-                """;
+
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/JsonSourceGeneratorTests.cs:766]:
```suggestion
```

File: src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/JsonSourceGeneratorTests.cs [hunk 8]
```diff
@@ -763,10 +763,38 @@ internal partial class JsonContext : JsonSerializerContext
                     {
                     }
                 }
-                """;
+                """
+            ;
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/JsonSourceGeneratorTests.cs:767]:
```suggestion
                """;
```

**Review by eiriktsarpalis — APPROVED** [review]
Thank you @manandre !

**Pull request merged**